core/arch/arch.c

```diff
@@ -1963,9 +1963,8 @@ get_ibl_routine_type_ex(dcontext_t *dcontext, cache_pc target, ibl_type_t *type
 
     /* a decent compiler should inline these nested loops */
     /* iterate in order <linked, unlinked> */
-    for (link_state = IBL_LINKED;
-         /* keep in mind we need a signed comparison when going downwards */
-         (int)link_state >= (int)IBL_UNLINKED; link_state-- ) {
```

Comment:
Wouldn't it be enough to change the condition in the for loop to `link_state != IBL_UNLINKED`?

Reply:
I spoke too soon

Reply:
Ok, it seems like the problem here is that casting from unsigned to signed is implementation defined, if the value is out-of-range for the signed type, right? So ` ibl_entry_point_type_t ` must be something like `unsigned int`. 

So we could change the cast to something like `int64_t`, that should be ok, as long as the compiler would not choose `uint64_t` for the enum. Or  `link_state` as `int` instead?

Reply:
I think declaring `link_state` as `int` would also do.  The aim here is to make the compiler choose the smallest possible size type for enums.

Reply:
Taking a step back, I'm not sure we thought through the API implications of -fshort-enums?  The option docs say:

```
Warning: the -fshort-enums switch causes GCC to generate code that is not binary compatible with code generated without that switch. Use it to conform to a non-default application binary interface. 
```

There are a number of enum types in the API headers.  I'm worried we'll end up with situations where DR passes an enum as a parameter and only initializes one byte, while the client code reads 4 bytes and gets garbage.

Reply:
I agree that changing the size of the enums would cause problems unless the client code is recompiled.  Do you think advertising this to the clients would be enough?

Reply:
I think regardless of whether we switch the enum type, I think it would still be worth fixing this, as the enum type is implementation defined and casting from unsigned to signed is undefined behaviour, for negative values AFAIK.

Reply:
Agreed, we can move the -fshort-enums discussion elsewhere.

Reply:
> Agreed, we can move the -fshort-enums discussion elsewhere.
A ticket, perhaps?

> I think declaring link_state as int would also do.
Personally I would prefer that, as it keeps the loop structure slightly more obvious and is less invasive. It should also be safe, as long as we do not have more enum values than INT_MAX. If we go for the break solution, please add a comment to the top of the loop, making it clear what the iteration space is and why it is this way.